File: Frontend/src/modules/admin/feature/stockMovement/services/stock-movement.service.ts
import api from "@/common/api";

import type {
  PageResponse,
  StockMovementDetailResponse,
  StockMovementListParams,
  StockMovementListResponse,
} from "../types/stock-movement.type";

const cleanParams = (
  params: Record<string, unknown>
): Record<string, unknown> =>
  Object.fromEntries(
    Object.entries(params).filter(
      ([, value]) =>
        value !== undefined &&
        value !== null &&
        value !== ""
    )
  );

const stockMovementService = {
  async getList(
    params: StockMovementListParams
  ): Promise<PageResponse<StockMovementListResponse>> {
    const response = await api.get(
      "/admin/stock-movements",
      {
        params: cleanParams({
          keyword: params.keyword?.trim() || undefined,
          inventoryLotId: params.inventoryLotId ?? undefined,
          movementType: params.movementType || undefined,
          createdBy: params.createdBy ?? undefined,
          referenceType: params.referenceType?.trim() || undefined,
          referenceId: params.referenceId ?? undefined,
          fromDate: params.fromDate || undefined,
          toDate: params.toDate || undefined,
          page: params.page ?? 0,
          size: params.size ?? 20,
        }),
      }
    );

    return response.data;
  },

  async getDetail(
    id: number
  ): Promise<StockMovementDetailResponse> {
    const response = await api.get(
      `/admin/stock-movements/${id}`
    );

    return response.data;
  },
};

export default stockMovementService;

File: Backend/src/main/java/org/example/datn_sd69/modules/stockmovement/service/StockMovementService.java
package org.example.datn_sd69.modules.stockmovement.service;

import org.example.datn_sd69.enums.StockMovementType;

import org.example.datn_sd69.modules.stockmovement.dto.response.StockMovementDetailResponse;
import org.example.datn_sd69.modules.stockmovement.dto.response.StockMovementListResponse;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;

public interface StockMovementService {

    Page<StockMovementListResponse> getList(

            String keyword,

            Integer inventoryLotId,

            StockMovementType movementType,

            Integer createdBy,

            String referenceType,

            Integer referenceId,

            LocalDate fromDate,

            LocalDate toDate,

            Pageable pageable
    );


    StockMovementDetailResponse getDetail(
            Integer id
    );
}